### PR TITLE
[dagster-sdf] Add Assets Decorator and Information Schema Reader

### PR DESCRIPTION
**Pull request by akbog**
## Summary & Motivation

This PR introduces the `sdf_assets` decorator and `SdfInformationSchema` Reader Class.

1. `sdf_assets` decorator

Wraps a `multi_asset` definition, which simply defines the workspace as a single op / graph. This can make sense in the context of an sdf workspace for several reasons (namely caching). The decorator takes as inputs:

  - target_dir: the target path to store sdf outputs in (i.e. materialized data from the sdf db, cache, etc.)
  - environment: which environment to run sdf commands against (default: dbg)

The decorator namely depends on the `SdfInformationSchema`.

2. `SdfInformationSchema` class

This is a wrapper class for interacting with sdf's compile time generated information schema. The information schema is a store of various metadata, such as classifiers, table materializations (i.e. table, view, external etc.), table and column level lineage, and table schema information. We load the information schema using pandas, extracting the dagster dag and schema information from the tables and columns dataframes respectively. These are both located in the sdf_target dir under `sdf_target/data/system/information_schema::sdf/`.

## How I Tested These Changes

See tests in `dagster_sdf_tests/test_asset_decorator.py` and `dagster_sdf_tests/test_information_schema.py`

### CODE REVIEW

File: python_modules/libraries/dagster-sdf/dagster_sdf/resource.py [hunk 1]
```diff
@@ -458,9 +473,10 @@ def cli(
 
         args = [
             self.sdf_executable,
-            *self.global_config_flags,
             *log_level_args,
+            *log_form_args,
             *args,
+            *self.global_config_flags,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/resource.py:479]: The point of global config flags is to configure the flags passed to `sdf` before a specific command is invoked. So this should come right after the executable.

It should essentially only accept these options:

```
❯ sdf --help          
SDF: A fast SQL compiler, local development framework, and in-memory analytical database

Usage: sdf [OPTIONS] <COMMAND>

Commands:
  new      Create a new sdf workspace
  clean    Remove artifacts that sdf has generated in the past
  compile  Compile models
  run      Run models
  test     Test your models
  stats    Statistics for your data
  report   Report code quality
  check    Check code quality
  lineage  Display lineage for a given table and/or column
  push     Push a local workspace to the SDF Service
  system   System maintenance, install and update
  auth     Authenticate CLI to services like SDF, AWS, OpenAI, etc
  man      Display reference material, like the CLI, dialect specific functions, schemas for authoring and interchange
  dbt      Initialize an sdf workspace from an existing dbt project
  help     Print this message or the help of the given subcommand(s)

Options:
      --log-level <LOG_LEVEL>
          Set log level
          
          [possible values: trace, debug, info, warn, error]

      --log-file <LOG_FILE>
          Creates or replaces the log file

      --log-form <LOG_FORM>
          Set the log format; default flat

          Possible values:
          - flat:   Writes nd-json, compact, no nested objects
          - nested: Writes nd-json, compact, nested objects
          - pretty: Writes nd-json, pretty, nested objects
```

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    @classmethod
+    def from_dict(cls, dict_data):
+        instance = cls()
+        for key, data in dict_data.items():
+            instance.data[key] = pd.DataFrame(data)
+            instance.loaded[key] = True
+        return instance
+
+
+class SdfInformationSchema(BaseModel):
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:70]: I'll refactor this later to use `from dagster._record import record`, so just a heads up.

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        # Convert to a serializable dictionary format
+        return {k: v.to_dict() for k, v in self.data.items() if self.loaded[k]}
+
+    @classmethod
+    def from_dict(cls, dict_data):
+        instance = cls()
+        for key, data in dict_data.items():
+            instance.data[key] = pd.DataFrame(data)
+            instance.loaded[key] = True
+        return instance
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:67]: not used?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    SDF_INFORMATION_SCHEMA_TABLES,
+    SDF_TARGET_DIR,
+)
+from .errors import DagsterSdfInformationSchemaNotFoundError
+
+
+class LazyDataFrameDict:
+    def __init__(
+        self,
+        information_schema_dir: Union[Path, str],
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:29]: This is an internally instantiated class, so can we stick with only `Path`?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
+import os
+from pathlib import Path
+from typing import Dict, List, Mapping, Optional, Set, Tuple, Union
+
+import pandas as pd
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:5]: nit: can we use polars instead?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+from .asset_utils import dagster_name_fn, default_asset_key_fn
+from .constants import (
+    DEFAULT_SDF_WORKSPACE_ENVIRONMENT,
+    SDF_INFORMATION_SCHEMA_TABLES,
+    SDF_TARGET_DIR,
+)
+from .errors import DagsterSdfInformationSchemaNotFoundError
+
+
+class LazyDataFrameDict:
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:26]: Could you write a brief docstring on why this is necessary? How large are these dataframes that we need to do this in a lazy fashion?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        workspace_dir: Union[Path, str],
+        target_dir: Union[Path, str],
+        environment: str = DEFAULT_SDF_WORKSPACE_ENVIRONMENT,
+        **data,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:84]: why does this accept kwargs?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    ):
+        workspace_dir = Path(workspace_dir) if isinstance(workspace_dir, str) else workspace_dir
+        target_dir = Path(target_dir) if isinstance(target_dir, str) else target_dir
+        information_schema_dir = target_dir.joinpath(
+            SDF_TARGET_DIR, environment, "data", "system", "information_schema::sdf"
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:89]: Why do these both need to be supplied? seems like we just need only one of them?

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py (hunk 2 above)
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:70]: Docstring please, along with any information about the format of the underlying information schema (e.g. what's available for access, what tables are available, etc)

File: python_modules/libraries/dagster-sdf/dagster_sdf/resource.py [hunk 9]
```diff
@@ -84,8 +86,17 @@ def to_default_asset_events(
             return
 
         table_id = self.raw_event["table"]
+        timestamp_str: str = self.raw_event["__ts"]
+        try:
+            dt = parser.parse(timestamp_str)
+            float_timestamp = dt.timestamp()
+            materialized_at = TimestampMetadataValue(float_timestamp)
+        except:
+            materialized_at = timestamp_str
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/resource.py:95]: we don't need this, as Dagster will report the materialization time by default

File: python_modules/libraries/dagster-sdf/dagster_sdf/resource.py [hunk 10]
```diff
@@ -423,7 +434,7 @@ def cli(
         args: Sequence[str],
         *,
         target_dir: Optional[Path] = None,
-        environment: str = DEFAULT_SDF_WORKSPACE_ENVIRONMENT,
+        environment: Optional[str] = None,
```
Comment (rexledesma) [inline python_modules/libraries/dagster-sdf/dagster_sdf/resource.py:437]: revert

File: python_modules/libraries/dagster-sdf/dagster_sdf/resource.py (hunk 1 above)
Comment (akbog) [reply python_modules/libraries/dagster-sdf/dagster_sdf/resource.py:479]: wires got crossed here 🤤 you're absolutely right. We should likely have `--show=none` as a global config as well but I'm checking internally why we haven't made that call

**Review by rexledesma — CHANGES_REQUESTED** [review]
requesting changes to put it back in your queue.

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py (hunk 6 above)
Comment (akbog) [reply python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py:26]: Dataframes should not be too large, but for large workspaces with complex dags they could be in the range of 50mb and over is my guess

File: python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py (hunk 5 above)
Comment (akbog) [reply python_modules/libraries/dagster-sdf/dagster_sdf/information_schema.py]: done!

**Review by rexledesma — APPROVED** [review]
I removed the changes related to adding metadata like code references, classifiers, column schema, and storage because there weren't any tests for them. Once there are tests, feel free to add them back one by one in separate PRs.

I also removed the tests around invoking the `lineage` project since it has a check, but the check's type doesn't seem to be correctly reflected in the information schema (its `purpose` was `model`). Let me know if there's a blessed way to determine if a table is a check so we can exclude those tables from asset dependencies.

I also simplified the API here since we didn't need the intermediate lazy dict class.

**Pull request merged**